src/hooks/ticker.ts

```diff
@@ -0,0 +1,164 @@
+import { useCallback, useEffect, useRef, useState } from 'react'
+
+// Only support specific token pairs until we find a method to validate token pairs
+export enum TokenPair {
+  BNBUSDT = 'bnbusdt',
+}
+
+/**
+ * @see https://binance-docs.github.io/apidocs/spot/en/#individual-symbol-ticker-streams
+ */
+export interface StreamData {
+  e: string
+  E: number
+  s: string
+  p: string
+  P: string
+  w: string
+  x: string
+  c: string
+  Q: string
+  b: string
+  B: string
+  a: string
+  A: string
+  o: string
+  h: string
+  l: string
+  v: string
+  q: string
+  O: number
+  C: number
+  F: number
+  L: number
+  n: number
+}
+
+export interface TickerStream {
+  eventType: string
+  eventTime: number
+  symbol: string
+  priceChange: number
+  priceChangePercent: number
+  weightAveragePrice: number
+  firstTrade: number
+  lastPrice: number
+  lastQuantity: number
+  bestBidPrice: number
+  bestBidQuantity: number
+  bestAskPrice: number
+  bestAskQuantity: number
+  openPrice: number
+  highPrice: number
+  lowPrice: number
+  totalTradedBaseAssetVolume: number
+  totalTradedQuoteAssetVolume: number
+  statisticsOpenTime: number
+  statisticsCloseTime: number
+  firstTradeId: number
+  lastTradeId: number
+  totalNumberOfTrades: number
+}
+
+export const useTokenPairTicker = (tokenPair: TokenPair, connectOnMount: boolean, retryCount = 3) => {
+  const retriesRemaining = useRef(retryCount)
+
+  // Chrome always returns an event code of 1006 so we have to keep track whether we closed the
+  // connection manually or not (1000)
+  const isManualClose = useRef(false)
+
+  // Use a ref instead of state so we mark the connection as closed immediately without
+  // triggering a re-render
+  const isConnected = useRef(false)
+  const [stream, setStream] = useState<TickerStream>(null)
+  const websocket = useRef<WebSocket>(null)
+
+  const connect = useCallback(() => {
+    if (navigator.onLine && !isConnected.current) {
+      const ws = new WebSocket(`wss://stream.binance.com:9443/ws/streams/${tokenPair}@ticker`)
```

Comment:
I wonder if disconnected WebSocket instances will pile up in memory when reconnecting. Sombody in stackoverflow comments also [was wondering about it](https://stackoverflow.com/questions/22431751/websocket-how-to-automatically-reconnect-after-it-dies#comment88921382_23176223) but nobody answered that.

My guess is once `ws.onclose()` handler finishes its job the memory gets cleaned, so it's probably safe to leave it like that.

Reply:
I don't think that is a problem here as the websocket instance gets overwritten each time connect is called and it can only do that if the previous once has closed. I am also not entirely sure...